### PR TITLE
Implementing theory solver for difference logic, both for Reals and Ints

### PR DESCRIPTION
**Pull request by blishko**
This PR add a specialized algorithm for solving difference constraints (difference logic in SMT-LIB terminology) over both Reals and Integers.

Credits for the implementation: @lunakv

### CODE REVIEW

**Comment by blishko** [comment]
There is still an issue regarding the computation of models in the presence of substitutions, but we should be able to implement a simple workaround. You can start the review in the meantime, @aehyvari .

File: src/logics/CMakeLists.txt [hunk 1]
```diff
@@ -10,6 +10,8 @@ target_sources(logics
 	PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/LRALogic.cc"
 	PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/LRALogic.h"
 	PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/LRATheory.cc"
+    PRIVATE "${CMAKE_CURRENT_SOURCE_DIR}/IDLTheory.cc"
```
Comment (aehyvari) [inline src/logics/CMakeLists.txt:13]: Check indentation

File: src/tsolvers/CMakeLists.txt [hunk 2]
```diff
@@ -29,8 +31,9 @@ include(lasolver/CMakeLists.txt)
 include(lrasolver/CMakeLists.txt)
 include(bvsolver/CMakeLists.txt)
 include(liasolver/CMakeLists.txt)
+include(stpsolver/CMakeLists.txt)
 
-install(FILES Deductions.h UFTHandler.h CUFTHandler.h LRATHandler.h LIATHandler.h TSolver.h THandler.h
+install(FILES Deductions.h UFTHandler.h CUFTHandler.h LRATHandler.h LIATHandler.h IDLTHandler.h TSolver.h THandler.h
```
Comment (aehyvari) [inline src/tsolvers/CMakeLists.txt:36]: Should there also be `RDLTHandler.h`?

File: src/tsolvers/stpsolver/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    PRIVATE "${CMAKE_CURRENT_LIST_DIR}/STPGraphManager.h"
+    PRIVATE "${CMAKE_CURRENT_LIST_DIR}/STPModel.h"
+    PRIVATE "${CMAKE_CURRENT_LIST_DIR}/SafeInt.h"
+    PRIVATE "${CMAKE_CURRENT_LIST_DIR}/Converter.h"
+
+    PRIVATE "${CMAKE_CURRENT_LIST_DIR}/STPSolver_implementations.hpp"
+    PRIVATE "${CMAKE_CURRENT_LIST_DIR}/STPStore_implementations.hpp"
+    PRIVATE "${CMAKE_CURRENT_LIST_DIR}/STPMapper_implementations.hpp"
+    PRIVATE "${CMAKE_CURRENT_LIST_DIR}/STPGraphManager_implementations.hpp"
+    PRIVATE "${CMAKE_CURRENT_LIST_DIR}/STPModel_implementations.hpp"
```
Comment (aehyvari) [inline src/tsolvers/stpsolver/CMakeLists.txt:15]: Change STPSolver_implementations.hpp => STPSolver.cc etc.?

File: src/tsolvers/stpsolver/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    PRIVATE "${CMAKE_CURRENT_LIST_DIR}/STPEdgeGraph.cpp"
```
Comment (aehyvari) [inline src/tsolvers/stpsolver/CMakeLists.txt:16]: Change to .cc?

File: src/tsolvers/stpsolver/STPSolver_implementations.hpp [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+    // Typically multiple backtrack points are popped together, hence popBacktackPoints is the important method to implement
+    popBacktrackPoints(1);
+}
+
+template<class T>
+void STPSolver<T>::popBacktrackPoints(unsigned int i) {
+    // This method is called after unsatisfiable state is detected
+    // The solver should remove all constraints that were pushed to the solver in the last "i" backtrackpoints
+    if (!i) return;
+    assert(backtrack_points.size() >= i);
```
Comment (aehyvari) [inline src/tsolvers/stpsolver/STPSolver_implementations.hpp:188]:
```suggestion
    assert(static_cast<unsigned int>(backtrack_points.size()) >= i);
```

**Review by aehyvari — COMMENTED** [review]
Here are some quick comments @blishko. I'll continue going through the code at a later time.

File: src/tsolvers/stpsolver/CMakeLists.txt (hunk 3 above)
Comment (blishko) [reply src/tsolvers/stpsolver/CMakeLists.txt:15]: The problem is that these files contain templated classes (templated over the Value type, which is different for RDL and for IDL) so they cannot be .cc files.

**Comment by lunakv** [comment]
For reference, the solver implements an algorithm described in [this paper](https://www.cs.upc.edu/~oliveras/espai/papers/cav05.pdf) based on exhaustive theory propagation.

**Comment by aehyvari** [comment]
Many thanks @lunakv and @jankofron!  We should have a meeting at some point to see if there's more to be done.

> For reference, the solver implements an algorithm described in this paper based on exhaustive theory propagation.

For generations to come it would be convenient to have this citation in the source code somewhere!

File: src/tsolvers/stpsolver/CMakeLists.txt (hunk 3 above)
Comment (aehyvari) [reply src/tsolvers/stpsolver/CMakeLists.txt:15]: Right, I remember having discussed this with you now!

File: src/tsolvers/CMakeLists.txt (hunk 2 above)
Comment (blishko) [reply src/tsolvers/CMakeLists.txt]: Fixed.

File: src/tsolvers/stpsolver/CMakeLists.txt (hunk 4 above)
Comment (blishko) [reply src/tsolvers/stpsolver/CMakeLists.txt]: Changed.

File: src/tsolvers/stpsolver/STPSolver_implementations.hpp (hunk 5 above)
Comment (blishko) [reply src/tsolvers/stpsolver/STPSolver_implementations.hpp]: Fixed.

File: src/tsolvers/stpsolver/STPStore.h [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+struct EdgeRef {
+    uint32_t x;
+
+    inline bool operator==(EdgeRef other) const { return x == other.x; }
+
+    inline bool operator!=(EdgeRef other) const { return x != other.x; }
+};
+
+static VertexRef VertRef_Undef = VertexRef{INT32_MAX};
```
Comment (aehyvari) [inline src/tsolvers/stpsolver/STPStore.h:31]: `UINT32_MAX` ?

File: src/tsolvers/stpsolver/STPStore.h [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+static EdgeRef EdgeRef_Undef = EdgeRef{INT32_MAX};
```
Comment (aehyvari) [inline src/tsolvers/stpsolver/STPStore.h:32]: Same here.

**Review by aehyvari — APPROVED** [review]
I left some very minor comments.  I think this is good now.

File: src/tsolvers/stpsolver/STPModel.h [hunk 8]
```diff
@@ -0,0 +1,42 @@
+#ifndef OPENSMT_STPVALMAPPER_HPP
+#define OPENSMT_STPVALMAPPER_HPP
+
+#include "STPGraphManager.h"
+
+#include <unordered_map>
+// implementations of template functions #included below class definition
+
+// holds the mapping from vertices to their values
+// FIXME: Incomplete for now. Will be fixed with new model interface
```
[the remark this answers is not in the document]
Comment (blishko) [reply src/tsolvers/stpsolver/STPModel.h:10]: No, this is now taken care of.

File: src/tsolvers/stpsolver/STPStore.h (hunk 6 above)
Comment (blishko) [reply src/tsolvers/stpsolver/STPStore.h:31]: Fixed.

File: src/tsolvers/stpsolver/STPSolver_implementations.hpp [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+    // This is just for compatibility with older architecture
+    // Typically multiple backtrack points are popped together, hence popBacktackPoints is the important method to implement
+    popBacktrackPoints(1);
+}
+
+template<class T>
+void STPSolver<T>::popBacktrackPoints(unsigned int i) {
+    // This method is called after unsatisfiable state is detected
+    // The solver should remove all constraints that were pushed to the solver in the last "i" backtrackpoints
+    if (!i) return;
```
[the remark this answers is not in the document]
Comment (blishko) [reply src/tsolvers/stpsolver/STPSolver_implementations.hpp:187]: Fixed.

**Pull request merged**